### ISSUE TITLE
Cannot Access Objects in Feature from Application(s)

### ISSUE THREAD

**Issue by alexclang** [OpenLiberty/open-liberty#24717]
I have written and debugged a Liberty feature which contains a large cache. It also has Java classes in it which act as a proxy (helper) for accessing that cache. I can't seem to access that helper class (contained in the feature) from any of my 20 or so applications (WARs, EARs) which require access to that cache. I get a NoClassDefFound exception. 

Is this a bug or feature? Any idea how I might achieve this goal?

**Comment by dmuelle** [comment]
Hi @alexclang - thanks for opening an issue. This repo is specifically for issues with our documentation. I'm transferring this issue over to our development repo where someone can help with your question.

**Comment by jhanders34** [comment]
This issue was also asked in a forum and the question asked was does your feature include that package as an API for your applications to use?

**Comment by Azquelt** [comment]
The header you would need to use in your feature manifest is [IBM-API-Package](https://www.ibm.com/docs/en/was-liberty/core?topic=manually-liberty-feature-manifest-files#rwlp_feat_definition__title__10)

**Comment by alexclang** [comment]
Yes, at the end of the day the object is on the feature. However, applications get a handle to it via JNDI via a context lookup. The result of the lookup is cast to the object type in the feature (the proxy class) which is what the application actually executes method calls on.

**Comment by alexclang** [comment]
To answer jhanders34, you have the polarity backwards. The application will get a JNDI handle to the object in the feature. The application (with the JNDI handle) will make a method call on a class in the feature which tries to do a class.ForName on a class in the application (what comes out of the feature object is a class name that the class in the feature tries to validate). This you have the classLoader in the feature trying to load a class that only exists in the application.

**Comment by alexclang** [comment]
I thought this would be useful information to the group: Yes, we do have the IBM-API-Package set for the packages we need to export from the feature (see in the ESA POM). But that just exposes the feature's classes to the EARs, what we need is the EAR's (and the common library used by the EARs) classes accessible to the feature.
